### PR TITLE
Si 569 enhanced set array

### PR DESCRIPTION
**Pull request by RasmusWind**
Rewrote parts of 'cmd_set' to handle setting parts of an array to an array of values.
Example usages:
Example param array: arr = {1, 2, 3, 4, 5, 6, 7, 8}
- (specific slice with value array) set arr[2:5] [9 9 9] -> {1, 2, 9, 9, 9, 6, 7, 8}
- (specific slice with single value) set arr[2:5] 9 -> {1, 2, 9, 9, 9, 6, 7, 8}
- (No slice with value array) set arr [5 4 5 4 5 4 5 4] -> {5, 4, 5, 4, 5, 4, 5, 4}
- (Negative index) set arr[-5:7] [1 2 3 4] -> {1, 2, 3, 1, 2, 3, 4, 8}
- (No start index) set arr[:5] [6 6 6 6] -> {6, 6, 6, 6, 2, 3, 4, 8}
- (No end index) set arr[5:] [6 6 6 6] -> {1, 2, 3, 4, 6, 6, 6, 6}
- (No index) set arr[:] [6 6 6 6 6 6 6 6] -> {6, 6, 6, 6, 6, 6, 6, 6}
The value array has to match the length of the slice. If it doesn't, then cmd_set will return SLASH_EINVAL.

Any feedback is very much appreciated.

### CODE REVIEW

File: src/param/param_slash.c [hunk 1]
```diff
@@ -30,6 +30,49 @@
...
+	char * saveptr;
+	char * token;
+	char slice_colon;
+	
+	strtok_r(arg, "[", &saveptr);
+	token = strtok_r(NULL, "[", &saveptr);
+	if (token != NULL) {
+        /* Search for the '[' symbol: */
+		// Searches for the format digit:digit.
+		sscanf(token, "%d%c%d", &offsets[0], &slice_colon, &offsets[1]);
```
Comment (Lykkeberg) [inline src/param/param_slash.c:46]: You might want to be a little defensive in handling the function arguments. Especially the ones which points to data to be "returned". The `offset` pointer and the `slice_detection` could be NULL. So could `param` and `arg`

File: src/param/param_slash.c [hunk 2]
```diff
@@ -30,6 +30,49 @@
...
+		// If the input was ":4" then no offsets will be set by the first sscanf, 
+		// so we check for this and try again with another format to match.
+		if (offsets[0] == INT_MIN && offsets[1] == INT_MIN){
+			sscanf(token, "%c%d", &slice_colon, &offsets[1]);
+		}
+		if(slice_colon == ':'){
+			*slice_detection = true;
```
Comment (Lykkeberg) [inline src/param/param_slash.c:53]: The `slice_detection` expects here that the caller has set it to `false` prior to calling the function which might not be a good semantics.

File: src/param/param_slash.c [hunk 3]
```diff
@@ -256,23 +340,96 @@ static int cmd_set(struct slash *slash) {
         optparse_del(parser);
 		return SLASH_EINVAL;
 	}
+	
+	// Create a queue, so that we can set the param in a single packet.
+	param_queue_t queue;
+	char queue_buf[PARAM_SERVER_MTU];
+	param_queue_init(&queue, queue_buf, PARAM_SERVER_MTU, 0, PARAM_QUEUE_TYPE_SET, 2);
```
Comment (Lykkeberg) [inline src/param/param_slash.c:347]: Could this fail, and render the `queue` invalid?

File: src/param/param_slash.c [hunk 4]
```diff
@@ -256,23 +340,96 @@ static int cmd_set(struct slash *slash) {
...
+		}
+		iterations++;
+
+		// If we hit the end value by detecting ']', and start index and end index are not equal, 
+		// then the slice length must be greater than the value array length.
+		// This shouldnt be possible, so we throw an error.
+		if(should_break == 0 && start_index != end_index){
+			optparse_del(parser);
+			fprintf(stderr, "Array slice is longer than value array\n");
+			return SLASH_EINVAL;
```
Comment (Lykkeberg) [inline src/param/param_slash.c:426]: What about the `queue`? Do you need to destroy it?

File: src/param/param_slash.c [hunk 5]
```diff
@@ -283,7 +440,7 @@ static int cmd_set(struct slash *slash) {
 		csp_timestamp_t time_now;
 		csp_clock_get_time(&time_now);
 		*param->timestamp = 0;
-		if (param_push_single(param, offset, valuebuf, 0, dest, slash_dfl_timeout, paramver, ack_with_pull) < 0) {
+		if (param_push_queue(&queue, 3, dest, slash_dfl_timeout, 0, ack_with_pull) < 0) {
```
Comment (Lykkeberg) [inline src/param/param_slash.c:443]: Shouldn't we use the `VERBOSE` defined value instead of 3?

File: src/param/param_slash.c [hunk 6]
```diff
@@ -30,6 +30,49 @@
...
+	
+	strtok_r(arg, "[", &saveptr);
+	token = strtok_r(NULL, "[", &saveptr);
+	if (token != NULL) {
+        /* Search for the '[' symbol: */
+		// Searches for the format digit:digit.
+		sscanf(token, "%d%c%d", &offsets[0], &slice_colon, &offsets[1]);
+		// If the input was ":4" then no offsets will be set by the first sscanf, 
+		// so we check for this and try again with another format to match.
+		if (offsets[0] == INT_MIN && offsets[1] == INT_MIN){
```
Comment (Lykkeberg) [inline src/param/param_slash.c:49]: Again the semantics on this. This function expects the caller to setup output arguments to known initial values prior to calling. This might be better adjusted in this method.

File: src/param/param_slash.c (hunk 3 above)
Comment (RasmusWind) [reply src/param/param_slash.c:357]: No, this cannot fail. By also looking at other implementations of queue, this seems to be the way to go.

File: src/param/param_slash.c (hunk 4 above)
Comment (RasmusWind) [reply src/param/param_slash.c:436]: Queue is free'd when we go out of scope. End of cmd_set.

File: src/param/param_slash.c [hunk 7]
```diff
@@ -30,6 +30,85 @@
...
+	if (id != ULONG_MAX && *endptr == '\0') {
+		*param = param_list_find_id(node, id);
+	} else {
+		*param = param_list_find_name(node, arg);
+	}
+
+	if(*param == NULL){
+		return PARAM_NOT_FOUND;
+	}
+
```
Comment (jeanbaptistelab) [inline src/param/param_slash.c:100]: I would move this part of the code out of this function and leave the responsibility of checking if the parsed parameter array is valid to the caller.
This would allow very easy unit testing of this function ;)

Comment (kivkiv12345) [reply src/param/param_slash.c:100]: Maybe it would be better if param_slash_parse_array() was split into 3 functions: 1 for business logic, 1 for error checking, and a public API that calls both? That way the public API wouldn't have to change (I realize param_slash_parse_array() hasn't been merged yet, but the error check is similar in param_slash_parse()).

File: src/param/param_string.c
Comment (jeanbaptistelab) [inline src/param/param_string.c:1]: Are you sure this change belongs to this PR ? It seems unrelated to "enhanced set array"...

Comment (RasmusWind) [reply src/param/param_string.c:1]: My mistake
